### PR TITLE
Waifu2X stability improvement

### LINKED ISSUE #425: [Bug]: Collapse crashes on startup
### Affected Version

Collapse 1.73.5 - Preview

## System Information
[DxDiag.txt](https://github.com/CollapseLauncher/Collapse/files/14560078/DxDiag.txt)

## Console Log
[log-2024-03-11-pre1.73.5-id10.log](https://github.com/CollapseLauncher/Collapse/files/14560093/log-2024-03-11-pre1.73.5-id10.log)

## Bug Behaviour
When attempting to launch Collapse, it launches a blank window which closes itself after a while.
Trying to debug what caused the crash, Console says about attempting to read/write protected memory, with error stack tracing to waifu2x function
A similar bug/issue is reported in the discord server, however I am able to launch the genshin executable without any issues

### Expected Behavior

Expected Behaviour:

Collapse works as intended.

### Steps to reproduce

1. Launch Collapse

### Related Issues

_No response_

### Screenshot(s)

Screenshot 2024-03-11 182148

### Additional Information

- Trying to launch collapse from \ProgramFiles\ or via shortcut had no differences
- when changing config.ini `EnableConsole=False` -> `EnableConsole=True` worked a few times, but didnt vice versa.

### PR DESCRIPTION
**Pull request by shatyuka**
Check list:
- [x] Unload Waifu2X if not enabled (get status only)
  - [x] Fix init multiple times
- [x] Check D3DMappingLayers and fallback to CPU mode. Fix #425
- [x] OpenMP blocktime ~https://github.com/Tencent/ncnn/pull/5370~ Use forked ncnn instead

### CODE REVIEW

File: CollapseLauncher/Classes/Helper/Image/ImageLoaderHelper.cs [hunk 1]
```diff
@@ -29,34 +29,63 @@ internal static class ImageLoaderHelper
...
+                        _cachedStatus = test.Status;
+                    }
+                }
+                return _cachedStatus;
+            }
+        }
 
         public static bool IsWaifu2XEnabled
         {
             get => GetAppConfigValue("EnableWaifu2X").ToBool() && IsWaifu2XUsable;
```
Comment (bagusnl) [inline CollapseLauncher/Classes/Helper/Image/ImageLoaderHelper.cs:54]: I think we don't need to check if waifu2x is usable if the config says its disabled

Comment (shatyuka) [reply CollapseLauncher/Classes/Helper/Image/ImageLoaderHelper.cs:54]: > I think we don't need to check if waifu2x is usable if the config says its disabled

a short-circuit evaluation happens here

Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Image/ImageLoaderHelper.cs:54]: after re reading the docs i finally get it 
:derp:

Thanks for the clarification

File: CollapseLauncher/Classes/Helper/Image/ImageLoaderHelper.cs [hunk 2]
```diff
@@ -29,34 +29,63 @@ internal static class ImageLoaderHelper
...
+        private static Waifu2X CreateWaifu2X()
+        {
+            var waifu2X = new Waifu2X();
+            if (waifu2X.Status < Waifu2XStatus.Error)
+            {
+                waifu2X.SetParam(Param.Noise, -1);
+                waifu2X.SetParam(Param.Scale, 2);
+                waifu2X.Load(Path.Combine(AppFolder!, @"Assets\Waifu2X_Models\scale2.0x_model.param.bin"),
+                    Path.Combine(AppFolder!, @"Assets\Waifu2X_Models\scale2.0x_model.bin"));
+            }
```
Comment (gablm) [inline CollapseLauncher/Classes/Helper/Image/ImageLoaderHelper.cs:74]: Saving the status here to the "cache" variable seems to solve the crash.

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs [hunk 3]
```diff
@@ -577,7 +576,10 @@ private string Waifu2XToolTip
                 var tooltip = Lang._SettingsPage.Waifu2X_Help;
                 if (ImageLoaderHelper.Waifu2XStatus == Waifu2XStatus.CpuMode)
                     tooltip += "\n\n" + Lang._SettingsPage.Waifu2X_Warning_CpuMode;
-                if (ImageLoaderHelper.Waifu2XStatus == Waifu2XStatus.NotAvailable)
```
Comment (gablm) [inline CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:580]: This would be more readable as a switch case

**Review by gablm — CHANGES_REQUESTED** [review]
Overall looks good apart Collapse crashing when Waifu2x is enabled and I try to disable it in the Settings page.

File: CollapseLauncher/Classes/Helper/Image/ImageLoaderHelper.cs (hunk 2 above)
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Image/ImageLoaderHelper.cs:75]: fixed in https://github.com/CollapseLauncher/Collapse/pull/428/commits/087d635bb67d16cd8ee99a91b56d245ea4b01e2e

File: CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs (hunk 3 above)
Comment (bagusnl) [reply CollapseLauncher/XAMLs/MainApp/Pages/SettingsPage.xaml.cs:580]: adjusted in https://github.com/CollapseLauncher/Collapse/pull/428/commits/71efeffd461ffae5a1bafa5be2bb23167dae36d6

**Review by bagusnl — APPROVED** [review]
approb

**Review by gablm — APPROVED** [review]
works as expected, looks good to me

**Review by neon-nyan — APPROVED** [review]
Approving the changes. Looking good to me!

**Pull request merged**